TopPlaces/TopPlaces/Controllers/DetailsPhotoViewController.m

```diff
@@ -0,0 +1,184 @@
+
+@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *spinner;
+@property (weak, nonatomic) IBOutlet UIScrollView *scrollView;
+@property (weak, nonatomic) IBOutlet UILabel *noImageLoadedView;
+@property (strong, nonatomic) UIImageView *imageView;
+
+@end
+
+@implementation DetailsPhotoViewController
+#pragma mark - UIView overrides
+
+- (void)viewDidLoad {
+  [super viewDidLoad];
+  self.scrollView.delegate = self;
+  
+  UITapGestureRecognizer *doubleTapRecognizer =
+      [[UITapGestureRecognizer alloc] initWithTarget:self
+                                              action:@selector(scrollViewDoubleFingerTapped:)];
+  
+  doubleTapRecognizer.numberOfTapsRequired = 2;
+  doubleTapRecognizer.numberOfTouchesRequired = 1;
+  [self.scrollView addGestureRecognizer:doubleTapRecognizer];
+
+  [self loadImage];
+  
+  NSLog(@"DetailsPhotoViewController::viewDidLoad self %p", self);
+}
+
+
+- (void)scrollViewDoubleFingerTapped:(UITapGestureRecognizer*)recognizer {
+  //if already zoomed out - zoom in a bit
+  if (self.scrollView.zoomScale == self.scrollView.minimumZoomScale) {
+    [self.scrollView setZoomScale:self.scrollView.zoomScale*2 animated:YES];
+  } else {
+    //zoom out
+    [self.scrollView setZoomScale:self.scrollView.minimumZoomScale animated:YES];
+  }
+}
+
+
+- (void)viewDidAppear:(BOOL)animated {
+  NSLog(@"DetailsPhotoViewController::viewDidAppear");
+  [super viewDidAppear:animated];
+}
+
+
+- (void)viewDidLayoutSubviews {
+  NSLog(@"DetailsPhotoViewController::viewDidLayoutSubviews");
+  [super viewDidLayoutSubviews];
+
+  if (self.imageView) {
+    [self aspectFitImage];
+  }
+}
+
+#pragma mark - UIScrollViewDelegate impl
+- (nullable UIView *)viewForZoomingInScrollView:(UIScrollView *)scrollView {
+  return self.imageView;
+}
+
+- (void)scrollViewDidZoom:(UIScrollView *)scrollView {
+  // The scroll view has zoomed, so you need to re-center the contents
+  [self centerScrollViewContents];
+}
+
+
+#pragma mark - aux logic
+- (void)loadImage {
+  //in ipad info can be missing
+  if (!self.photoInfo) {
+    return;
+  }
+
+  [PhotosHistory addPhotoInfo:self.photoInfo];
+  
+  self.noImageLoadedView.hidden = YES;
+  [self.spinner startAnimating];
+
+  __weak DetailsPhotoViewController *weakSelf = self;
+  dispatch_queue_t fetchPhoto = dispatch_queue_create("picture of photo", NULL);
+  dispatch_async(fetchPhoto, ^(void){
+    //get the image
+    UIImage *img = [weakSelf.placesPhotosProvider downloadPhoto:weakSelf.photoInfo];
+    
+    //show it (on UI thread)
+    dispatch_async(dispatch_get_main_queue(), ^(void){
+      [weakSelf.spinner stopAnimating];
+      
+      //if download failed
+      if (!img) {
+        weakSelf.noImageLoadedView.hidden = NO;
+        return;
+      }
+
+      weakSelf.title = weakSelf.photoInfo.title;
+      
+      weakSelf.imageView = [[UIImageView alloc] initWithImage:img];
+      weakSelf.imageView.frame = CGRectMake(0, 0, img.size.width, img.size.height);
```

Comment:
This is not needed, <code>[UIImageView initWithImage:]</code> sets the frame of the view according to the size of the image. It also take the screen scale factor (dpi) into account (eg. if the image size is 10x10 and screen scale is 2 it will set the frame size to 5x5 so it will be exactly 10x10 pixels).